xapian-applications/omega/query.cc

```diff
@@ -752,11 +756,13 @@ html_highlight(const string &s, const string &list,
 	}
 	j = term_end;
 	term = Xapian::Unicode::tolower(term);
-	int match = word_in_list(term, list);
+	build_word_map(list);
```

Comment:
That would achieve calling it exactly once, though not the reuse of the existing mapping when called again with the same list of terms.

C++ is an OO language, so the clean approach is to wrap the data (i.e. the mapping) as a private member of a new class, and make the functions which operate on it (such as `build_word_map()` and `word_in_list()`) member functions of that class.  The class could then also keep track of the list of terms which was used to build the current mapping, then `build_word_map()` could just return if passed the same list of terms as was used the last time.